File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.js
function solution(s) {
    const stack = [];

    for (const char of s) {
        if (char === '(') {
            stack.push(char);
        } else if (char === ')') {
            if (stack.length === 0) return false;  // 닫는 괄호인데 열림이 없음
            stack.pop();  // 올바른 짝
        }
    }

    return stack.length === 0;  // 다 짝지어졌는지 확인
}

File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.java
import java.util.Deque;
import java.util.ArrayDeque;

class Solution {
    boolean solution(String s) {
        Deque<Character> stack = new ArrayDeque<>();
        
        for (char ch: s.toCharArray()){
            if (ch == '('){
                stack.push(ch);
            } else if (ch == ')'){
                if (stack.isEmpty()){
                    return false;
                }
                stack.pop();
            }
        }
        
        return stack.isEmpty();
    }
}